rust/suseconnect-agama/src/lib.rs

```diff
@@ -0,0 +1,793 @@
+    MissingCredentialsFile(String),
+    #[error("JSON error: {0}")]
+    JSONError(String),
+    #[error("Network error: {0}")]
+    NetError(String),
+    #[error("Timeout: {0}")]
+    Timeout(String),
+    // TODO: check how it will look like if code display message won't be duplicite to connect message
+    #[error("SSL error: {message} (code: {code})")]
+    SSLError {
+        message: String,
+        code: SSLErrorCode,
+        current_certificate: String,
+    },
+    #[error(transparent)]
+    UTF8Error(#[from] std::ffi::IntoStringError),
+}
+
+/// checks response from suseconnect for errors
+///
+/// ruby counterpart is at https://github.com/SUSE/connect-ng/blob/main/third_party/yast/lib/suse/toolkit/shim_utils.rb#L32
+fn check_error(response: &Value) -> Result<(), Error> {
+    if let Some(error) = response.get("err_type") {
+        let Some(error_str) = error.as_str() else {
+            return Err(Error::UnexpectedResponse(response.to_string()));
+        };
+        let message = response
+            .get("message")
+            .and_then(|i| i.as_str())
+            .unwrap_or("No message")
+            .to_string();
+
+        match error_str {
+            "APIError" => {
+                let code = response.get("code").and_then(|i| i.as_u64()).unwrap_or(400);
+                return Err(Error::ApiError { message, code });
+            }
+            "MalformedSccCredentialsFile" => {
+                return Err(Error::MalformedSccCredentialsFile(message))
+            }
+            "MissingCredentialsFile" => return Err(Error::MissingCredentialsFile(message)),
+            "JSONError" => return Err(Error::JSONError(message)),
+            "NetError" => return Err(Error::NetError(message)),
+            "Timeout" => return Err(Error::Timeout(message)),
+            "SSLError" => {
+                let code = response.get("code").and_then(|i| i.as_u64()).unwrap_or(0);
+                let current_certificate = response
+                    .get("data")
+                    .and_then(|i| i.as_str())
+                    .unwrap_or("")
+                    .to_string();
+                return Err(Error::SSLError {
+                    message,
+                    code: SSLErrorCode::from_u64(code),
+                    current_certificate,
+                });
+            }
+            _ => {
+                return Err(Error::UknownError(response.to_string()));
+            }
+        }
+    }
+    Ok(())
+}
+
+/// SCC/System credentails
+///
+/// Data returned from announce call at https://github.com/SUSE/connect-ng/blob/main/third_party/yast/lib/suse/connect/yast.rb#L57
+#[derive(Debug, Clone)]
+pub struct Credentials {
+    pub login: String,
+    pub password: String,
+}
+
+impl TryFrom<Value> for Credentials {
+    type Error = Error;
+
+    fn try_from(value: Value) -> Result<Self, Self::Error> {
+        let Some(credentials) = value.get("credentials").and_then(|c| c.as_array()) else {
```

Comment:
I think you can do (just a suggestion):


```suggestion
        let Some(credentials) = value.get("credentials").and_then(Value::as_array) else {
```

You can do the same with others `as_*` (e.g., `as_str`).